### PR TITLE
Add support for aws_batch_compute_environment

### PR DESCRIPTION
**Pull request by shibataka000**
## Description
Add support for `aws_batch_compute_environment` to creates a AWS Batch compute environment.

## Tests
```
$ make testacc TESTARGS='-run=TestAccAWSBatchComputeEnvironment'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccAWSBatchComputeEnvironment -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSBatchComputeEnvironment_createEc2
--- PASS: TestAccAWSBatchComputeEnvironment_createEc2 (53.54s)
=== RUN   TestAccAWSBatchComputeEnvironment_createEc2WithTags
--- PASS: TestAccAWSBatchComputeEnvironment_createEc2WithTags (40.32s)
=== RUN   TestAccAWSBatchComputeEnvironment_createSpot
--- PASS: TestAccAWSBatchComputeEnvironment_createSpot (44.21s)
=== RUN   TestAccAWSBatchComputeEnvironment_createUnmanaged
--- PASS: TestAccAWSBatchComputeEnvironment_createUnmanaged (51.36s)
=== RUN   TestAccAWSBatchComputeEnvironment_updateMaxvCpus
--- PASS: TestAccAWSBatchComputeEnvironment_updateMaxvCpus (74.04s)
=== RUN   TestAccAWSBatchComputeEnvironment_updateInstanceType
--- PASS: TestAccAWSBatchComputeEnvironment_updateInstanceType (91.76s)
=== RUN   TestAccAWSBatchComputeEnvironment_updateComputeEnvironmentName
--- PASS: TestAccAWSBatchComputeEnvironment_updateComputeEnvironmentName (78.01s)
=== RUN   TestAccAWSBatchComputeEnvironment_createEc2WithoutComputeResources
--- PASS: TestAccAWSBatchComputeEnvironment_createEc2WithoutComputeResources (22.59s)
=== RUN   TestAccAWSBatchComputeEnvironment_createUnmanagedWithComputeResources
--- PASS: TestAccAWSBatchComputeEnvironment_createUnmanagedWithComputeResources (49.45s)
=== RUN   TestAccAWSBatchComputeEnvironment_createSpotWithoutBidPercentage
--- PASS: TestAccAWSBatchComputeEnvironment_createSpotWithoutBidPercentage (25.54s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	530.836s
```

### CODE REVIEW

**Comment by ashinohara** [comment]
Hello @shibataka000 are you also working on support for `aws_batch_job_definition` and `aws_batch_job_queue`?  I am asking because I have an implementation of those resources that also includes `aws_batch_compute_environment` but it seems you have beaten me to the punch.  I am happy to collaborate and provide my code for `aws_batch_job_definition` and `aws_batch_job_queue`.

**Comment by shibataka000** [comment]
Hello @ashinohara .
I have not been working on support for `aws_batch_job_definition` and `aws_batch_job_queue` yet.
So I am happy if you create PullRequest for supporting for them after this PullRequrest will be merged or closed.
Thank you.

**Comment by ashinohara** [comment]
Sounds good to me @shibataka000.  I will utilize your code for `aws_batch_compute_environment` once this gets merged and then submit my pull request for the 2 additional resources.

**Comment by mathewpeterson** [comment]
@shibataka000 Would you mind resolving the merge conflicts?

**Comment by shibataka000** [comment]
@mathewpeterson I resolved the merge conflicts. Thank you.

**Comment by mathewpeterson** [comment]
@radeksimko Is this PR something you can review? We currently have a need for this and have to rely on breaking out of terraform with local-exec to run python scripts which has it's own set of issues. It also sounds like @ashinohara has functioning code for `aws_batch_job_definition` and `aws_batch_job_queue` so we could have a fairly complete solution around AWS Batch.

**Comment by Ninir** [comment]
Hey folks,

For a better review, @shibataka000 do you think you could open a PR for the vendoring of Batch Compute vendors, and keep this one for the resources implementation?

Will then review this as soon as it is done! (Smaller PRs are always of a good help for us 😄 )

Thanks for the effort there!

**Comment by shibataka000** [comment]
@Ninir Thank you your advice.
I open a PR https://github.com/terraform-providers/terraform-provider-aws/pull/1457 for the vendoring of Batch Compute vendors.

**Comment by Ninir** [comment]
@shibataka000 Could you resolve conflicts here please?

Will make the review right after that! :)

**Comment by shibataka000** [comment]
@Ninir I resolved conflicts. Thank you.

**Comment by mathewpeterson** [comment]
@ashinohara Are you able to get your PR's ready for `aws_batch_job_definition` and `aws_batch_job_queue` resources?

**Comment by ashinohara** [comment]
@mathewpeterson Yes, I can do that.  Would you prefer that I branch from this branch since the `aws_batch_job_definition` and `aws_batch_job_queue` are dependent upon the `aws_batch_compute_environment` or should I wait until this PR is merged and then branch from master?

**Comment by mathewpeterson** [comment]
@ashinohara I _think_ you could open a PR against master and github should be able to merge cleanly.

**Comment by ashinohara** [comment]
@mathewpeterson Unfortunately my tests for `aws_batch_job_queue` require creating a `aws_batch_compute_environment` so in order for me to run those tests I need this feature to be available.  I currently have my own code for `aws_batch_compute_environment` but I would like to be able to test it with the code in this PR.

**Comment by Ninir** [comment]
Hey Folks,

@shibataka000 Thanks!
@ashinohara If you can, please wait for this review 😄 . 

Will make the review in the coming days! :)

**Comment by mathewpeterson** [comment]
@ashinohara Yes, sorry for the confusion.

I was just trying lead to the question: Have you updated your code use shibataka000's `aws_batch_compute_environment` implementation? 

Sorry for any confusion. I think it's best to listen to @Ninir and just wait for this PR to be reviewed.

File: aws/resource_aws_batch_compute_environment.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+				},
+			},
+			"service_role": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"state": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice([]string{"ENABLED", "DISABLED"}, true),
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Can you use constants in place of ENABLED & DISABLED?

File: aws/resource_aws_batch_compute_environment.go [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+				Default:      "ENABLED",
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"MANAGED", "UNMANAGED"}, true),
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Same here: could you use constants?

File: aws/resource_aws_batch_compute_environment.go [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+						"instance_type": {
+							Type:     schema.TypeSet,
+							Required: true,
+							ForceNew: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: You can omit this as far as I know!

File: aws/resource_aws_batch_compute_environment.go [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+						},
+						"tags": tagsSchema(),
+						"type": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ForceNew:     true,
+							ValidateFunc: validation.StringInSlice([]string{"EC2", "SPOT"}, true),
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Can you use constants here?

File: aws/resource_aws_batch_compute_environment.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+		},
+	}
+}
+
+func resourceAwsBatchComputeEnvironmentCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).batchconn
+
+	computeEnvironmentName := d.Get("compute_environment_name").(string)
+
+	log.Printf("[DEBUG] Create compute environment \"%s\".\n", computeEnvironmentName)
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: To get a better idea about what is transiting, can you move this line 214, and reword it to:
`log.Printf("[DEBUG] Creating Batch Compute Environment: %s", input)`?

Since the `CreateComputeEnvironmentInput` struct has a `String` method, it will display it nicely, and provide much more information when debugging!

File: aws/resource_aws_batch_compute_environment.go [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+		ComputeEnvironmentName: aws.String(computeEnvironmentName),
+		ServiceRole:            aws.String(serviceRole),
+		Type:                   aws.String(computeEnvironmentType),
+	}
+
+	if v, ok := d.GetOk("state"); ok {
+		input.State = aws.String(v.(string))
+	}
+
+	if computeEnvironmentType == "MANAGED" {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Can you check against the constant?

File: aws/resource_aws_batch_compute_environment.go [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+	}
+
+	result, err := conn.DescribeComputeEnvironments(input)
+	if err != nil {
+		return err
+	}
+
+	if len(result.ComputeEnvironments) == 0 {
+		return fmt.Errorf("One compute environment is expected, but AWS return no compute environment")
+	} else if len(result.ComputeEnvironments) >= 2 {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: This should be on a newline, since the return instruction line 245 returns
Also, is it possible to have 2 ComputeEnvironments with the same name?

File: aws/resource_aws_batch_compute_environment.go [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+
+	computeEnvironmentName := d.Get("compute_environment_name").(string)
+
+	input := &batch.DescribeComputeEnvironmentsInput{
+		ComputeEnvironments: []*string{
+			aws.String(computeEnvironmentName),
+		},
+	}
+
+	result, err := conn.DescribeComputeEnvironments(input)
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Could you write to the logs right before this line that we are actually reading the Batch Compute Environment, in the same way as the Create one?

File: aws/resource_aws_batch_compute_environment.go [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+	if err != nil {
+		return err
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Here, we should check that the environment still exists, i.e. it has not been removed from the console, and otherwise return the error

File: aws/resource_aws_batch_compute_environment.go [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+
+	return nil
+}
+
+func resourceAwsBatchComputeEnvironmentDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).batchconn
+
+	computeEnvironmentName := d.Get("compute_environment_name").(string)
+
+	log.Printf("[DEBUG] Delete compute environment \"%s\".\n", computeEnvironmentName)
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Could you put it line 296, in the same way that we did for create & read, in order to provide better debug?

File: aws/resource_aws_batch_compute_environment.go [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+
+	updateInput := &batch.UpdateComputeEnvironmentInput{
+		ComputeEnvironment: aws.String(computeEnvironmentName),
+		State:              aws.String("DISABLED"),
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: We should also use the constant here

File: aws/resource_aws_batch_compute_environment.go [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+
+	return nil
+}
+
+func resourceAwsBatchComputeEnvironmentUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).batchconn
+
+	computeEnvironmentName := d.Get("compute_environment_name").(string)
+
+	log.Printf("[DEBUG] Update compute environment \"%s\".\n", computeEnvironmentName)
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Could you put it line 364, in the same way that we did for create, read, delete, in order to provide better debug?

Thus, we could remove logs in `HasChange` blocks

File: aws/resource_aws_batch_compute_environment.go [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+
+	if d.HasChange("compute_resources") {
+		log.Printf("[DEBUG] Compute resources of \"%s\" has change.\n", computeEnvironmentName)
+		computeResources := d.Get("compute_resources").([]interface{})
+		if len(computeResources) == 0 {
+			return fmt.Errorf("One compute environment is expected, but no compute environments are set")
+		}
+		computeResource := computeResources[0].(map[string]interface{})
+
+		if d.HasChange("compute_resources.0.desired_vcpus") {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Looking at the HasChange code, I'm not sure we even need to check for children: we could simply pass options, since we know that `compute_resources ` changed. Thougths?

File: aws/resource_aws_batch_compute_environment.go [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+			return err
+		}
+
+		if len(result.ComputeEnvironments) == 0 {
+			return fmt.Errorf("One compute environment is expected, but AWS return no compute environment")
+		} else if len(result.ComputeEnvironments) >= 2 {
+			return fmt.Errorf("One compute environment is expected, but AWS return too many compute environment")
+		}
+		computeEnvironment := result.ComputeEnvironments[0]
+		if *(computeEnvironment.Status) == "VALID" {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Could you use the constant here?

File: aws/resource_aws_batch_compute_environment.go [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+				aws.String(computeEnvironmentName),
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		if len(result.ComputeEnvironments) == 0 {
+			return fmt.Errorf("One compute environment is expected, but AWS return no compute environment")
+		} else if len(result.ComputeEnvironments) >= 2 {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: This should be on a newline, since the previous line returns

File: aws/resource_aws_batch_compute_environment.go [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+			return fmt.Errorf("One compute environment is expected, but AWS return too many compute environment")
+		}
+		computeEnvironment := result.ComputeEnvironments[0]
+		if *(computeEnvironment.Status) == "VALID" {
+			break
+		} else {
+			time.Sleep(1 * time.Second)
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Not sure 1 second is enough... perhaps doing it every 5 seconds would be better.
By the way, do you think it would be possible to use [such behaviour](https://github.com/terraform-providers/terraform-provider-aws/blob/b8d4e1570fc43a2acee6b6e47f63c9db6b067fa2/aws/resource_aws_alb.go#L192-L216) instead of making this function?

(same would go for the deletion)

File: website/docs/r/batch_compute_environment.html.markdown [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+  }
+  service_role = "${aws_iam_role.aws_batch_service_role.arn}"
+  type = "MANAGED"
+}
+```
+
+## Argument Reference
+
+* `compute_environment_name` - (Required) The name for your compute environment. Up to 128 letters (uppercase and lowercase), numbers, and underscores are allowed.
+* `compute_resources` - (Optional) Details of the compute resources managed by the compute environment. This parameter is required for managed compute environments.
```
Comment (Ninir) [inline website/docs/r/batch_compute_environment.html.markdown]: Could you add an indication that the documentation is below?

File: website/docs/r/batch_compute_environment.html.markdown [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+* `service_role` - (Required) The full Amazon Resource Name (ARN) of the IAM role that allows AWS Batch to make calls to other AWS services on your behalf.
+* `state` - (Optional) The state of the compute environment. If the state is `ENABLED`, then the compute environment accepts jobs from a queue and can scale out automatically based on queues. Valid items are `ENABLED` or `DISABLED`. Defaults to `ENABLED`.
+* `type` - (Required) The type of the compute environment. Valid items are `MANAGED` or `UNMANAGED`.
+
+**compute_resources** is a child block with a single argument:
+
+* `bid_percentage` - (Optional) The minimum percentage that a Spot Instance price must be when compared with the On-Demand price for that instance type before instances are launched. 
+  For example, if your bid percentage is 20%, then the Spot price must be below 20% of the current On-Demand price for that EC2 instance.
+  This parameter is required for SPOT compute environments.
```
Comment (Ninir) [inline website/docs/r/batch_compute_environment.html.markdown]: I would put this line 116 too

File: website/docs/r/batch_compute_environment.html.markdown [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+* `desired_vcpus` - (Optional) The desired number of EC2 vCPUS in the compute environment.
+* `ec2_key_pair` - (Optional) The EC2 key pair that is used for instances launched in the compute environment.
+* `image_id` - (Optional) The Amazon Machine Image (AMI) ID used for instances launched in the compute environment.
+* `instance_role` - (Required) The Amazon ECS instance role applied to Amazon EC2 instances in a compute environment.
+* `instance_type` - (Required) The instances types that may launched.
```
Comment (Ninir) [inline website/docs/r/batch_compute_environment.html.markdown]: What do you think of `A list of instance types that may be launched`?

File: website/docs/r/batch_compute_environment.html.markdown [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+* `max_vcpus` - (Required) The maximum number of EC2 vCPUs that an environment can reach.
+* `min_vcpus` - (Required) The minimum number of EC2 vCPUs that an environment should maintain.
+* `security_group_ids` - (Required) The EC2 security group that is associated with instances launched in the compute environment.
```
Comment (Ninir) [inline website/docs/r/batch_compute_environment.html.markdown]: What do you think of `A list of EC2 security group IDS that are associated with instances...`?

File: website/docs/r/batch_compute_environment.html.markdown [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+* `spot_iam_fleet_role` - (Optional) The Amazon Resource Name (ARN) of the Amazon EC2 Spot Fleet IAM role applied to a SPOT compute environment.
+  This parameter is required for SPOT compute environments.
```
Comment (Ninir) [inline website/docs/r/batch_compute_environment.html.markdown]: Could you put this line 126?

File: website/docs/r/batch_compute_environment.html.markdown [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+* `subnets` - (Required) The VPC subnets into which the compute resources are launched.
```
Comment (Ninir) [inline website/docs/r/batch_compute_environment.html.markdown]: What do you think of `A list of VPC subnets into which the compute resources are launched.`?

**Comment by shibataka000** [comment]
Thank you for your advice. I'm updating codes.

File: aws/resource_aws_batch_compute_environment.go (hunk 9 above)
Comment (shibataka000) [reply aws/resource_aws_batch_compute_environment.go]: @Ninir Do you mean that we should check `err != nil` caused by no compute environment existing or other trouble?

As far as I know, if there are no computer environment, `DescribeComputeEnvironments` return `nil` as `err`.
Instead of it, length of array `result.ComputeEnvironment` is 0.
I have already checked it at [L275](https://github.com/terraform-providers/terraform-provider-aws/pull/1048/files#diff-14bfc58fb904c87239404e4628f71c44R275).

So I think it is not necessary to check whether compute environment exists or not at [L272](https://github.com/terraform-providers/terraform-provider-aws/pull/1048#discussion-diff-134530830R241).
What do you think?

If I have some misunderstanding, please tell me it.

**Comment by shibataka000** [comment]
@Ninir Thank you for your advice! I have fixed codes and documents except [this](https://github.com/terraform-providers/terraform-provider-aws/pull/1048#discussion_r134530830).
If there are any other advice, please give me it.

File: aws/resource_aws_batch_compute_environment.go [hunk 23]
```diff
@@ -0,0 +1,496 @@
...
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/batch"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+const (
+	MANAGED   = "MANAGED"
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Sorry if my comment was misleading, I meant to prefer the use of constants from the SDK, as exposed [here](https://github.com/terraform-providers/terraform-provider-aws/blob/master/vendor/github.com/aws/aws-sdk-go/service/batch/api.go#L4586)

File: aws/resource_aws_batch_compute_environment.go [hunk 24]
```diff
@@ -0,0 +1,496 @@
...
+		if len(result.ComputeEnvironments) == 0 {
+			return result, DELETED, nil
+		}
+
+		computeEnvironment := result.ComputeEnvironments[0]
+		return result, *(computeEnvironment.Status), nil
+	}
+}
+
+func isCorrentComputeEnvironmentName(i interface{}, k string) (s []string, es []error) {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: I think there is a typo in Corrent 😄 
Also, it would be better to put this function in validators.go, and rename it to `validateBatchComputeEnvironmentName`

File: website/aws.erb [hunk 25]
```diff
@@ -270,6 +270,15 @@
                     </ul>
                 </li>
 
+                <li<%= sidebar_current("docs-aws-resource-batch-compute-environment") %>>
```
Comment (Ninir) [inline website/aws.erb]: This should be `docs-aws-resource-batch`

File: aws/resource_aws_batch_compute_environment.go [hunk 26]
```diff
@@ -0,0 +1,496 @@
...
+
+	if !(reComputeEnvironmentName.MatchString(v) && len(v) <= 128) {
+		es = append(es, fmt.Errorf("computeEnvironmentName must be up to 128 letters (uppercase and lowercase), numbers, and underscores."))
+		return
+	}
+
+	return
+}
+
+func isArnOfIamRole(i interface{}, k string) (s []string, es []error) {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Same as above, it would be better to put this function in validators.go, and rename it to `validateBatchComputeEnvironmentIamRole`

File: aws/resource_aws_batch_compute_environment.go [hunk 27]
```diff
@@ -0,0 +1,496 @@
...
+	if !reArnOfIamRole.MatchString(v) {
+		es = append(es, fmt.Errorf("expected arn of iam role, got %s", v))
+		return
+
+	}
+
+	return
+}
+
+func isArnOfIamInstanceProfile(i interface{}, k string) (s []string, es []error) {
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Here also!

File: aws/resource_aws_batch_compute_environment.go [hunk 28]
```diff
@@ -0,0 +1,496 @@
...
+							Type:     schema.TypeSet,
+							Required: true,
+							ForceNew: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"spot_iam_fleet_role": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ForceNew:     true,
+							ValidateFunc: isArnOfIamRole,
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: It would be better to use the validateArn function, which has more guards to validate ARNs. Thoughts?

File: aws/resource_aws_batch_compute_environment.go [hunk 29]
```diff
@@ -0,0 +1,496 @@
...
+							ForceNew:     true,
+							ValidateFunc: validation.StringInSlice([]string{EC2, SPOT}, true),
+						},
+					},
+				},
+			},
+			"service_role": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: isArnOfIamRole,
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: It would be better to use the validateArn function, which has more guards to validate ARNs. Thoughts?

File: website/docs/r/batch_compute_environment.html.markdown [hunk 30]
```diff
@@ -0,0 +1,137 @@
+---
+layout: "aws"
+page_title: "AWS: aws_batch_compute_environment"
+sidebar_current: "docs-aws-resource-batch-compute-environment"
+description: |-
+  Creates a AWS Batch compute environment.
+---
+
+# aws\_batch\_compute\_environment
```
Comment (Ninir) [inline website/docs/r/batch_compute_environment.html.markdown]: Those `\` are not needed anymore, it is safe to remove them

**Comment by shibataka000** [comment]
@Ninir Thank you for your advice! I really appreciate your help!
If there are any other advice, please give me it.

File: aws/resource_aws_batch_compute_environment.go [hunk 31]
```diff
@@ -0,0 +1,429 @@
...
+				"max_vcpus":           computeResource.MaxvCpus,
+				"min_vcpus":           computeResource.MinvCpus,
+				"security_group_ids":  schema.NewSet(schema.HashString, flattenStringList(computeResource.SecurityGroupIds)),
+				"spot_iam_fleet_role": computeResource.SpotIamFleetRole,
+				"subnets":             schema.NewSet(schema.HashString, flattenStringList(computeResource.Subnets)),
+				"tags":                tagsToMapGeneric(computeResource.Tags),
+				"type":                computeResource.Type,
+			},
+		})
+	}
```
Comment (Ninir) [inline aws/resource_aws_batch_compute_environment.go]: Last thing to do would be to move this to a flatten function I think :)
It would then allow us to manage optional fields more easily, and be consistent with other parts of the code.

Once done, 'will run the tests and merge if it's all ok :)

**Comment by shibataka000** [comment]
@Ninir Thank you for your advice!

> Last thing to do would be to move this to a flatten function I think :)

For example, You mean [resource_aws_s3_bucket_notification.go#L402](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_s3_bucket_notification.go#L402)?
I think I did it. If I have some misunderstanding, please tell me that.

**Comment by shibataka000** [comment]
@Ninir I really appreciate your help!

@ashinohara  I'm sorry to have kept you waiting long time.
You can create PR for supporting for `aws_batch_job_definition` and `aws_batch_job_queue` .
Thank you.

**Comment by ashinohara** [comment]
No problem @shibataka000, I appreciate your contribution.  I am updating my code to work with yours and hope to submit a PR soon.
